src/blocks/service_status.rs

```diff
@@ -0,0 +1,143 @@
+//! Display the status of a service
+//!
+//! Right now only `systemd` is supported.
+//!
+//! # Configuration
+//!
+//! Key | Values | Default
+//! ----|--------|--------
+//! `driver` | Which init system is running the service. Available drivers are: `"systemd"` | `"systemd"`
+//! `service` | The name of the service | **Required**
+//! `format` | A string to customise the output of this block. See below for available placeholders. | `" $service $active_state "`
+//!
+//! Placeholder    | Value                     | Type   | Unit
+//! ---------------|---------------------------|--------|-----
+//! `service`      | The name of the service   | Text   | -
+//! `active_state` | The status of the service | Text   | -
+//!
+//! # Example
+//!
+//! ```toml
+//! [[block]]
+//! block = "service_status"
+//! service = "cups"
+//! ```
+//!
+
+use super::prelude::*;
+use zbus::dbus_proxy;
+use zbus::PropertyStream;
+
+#[derive(Deserialize, Debug, Default)]
+#[serde(default)]
+pub struct Config {
+    driver: DriverType,
+    service: String,
+    format: FormatConfig,
+}
+
+#[derive(Deserialize, Debug, SmartDefault)]
+#[serde(rename_all = "snake_case")]
+enum DriverType {
+    #[default]
+    Systemd,
+}
+
+pub async fn run(config: Config, mut api: CommonApi) -> Result<()> {
+    api.event_receiver.close();
+
+    let mut widget =
+        Widget::new().with_format(config.format.with_default(" $service $active_state ")?);
+
+    let mut driver: Box<dyn Driver> = match config.driver {
+        DriverType::Systemd => Box::new(SystemdDriver::new(config.service.clone()).await?),
+    };
+
+    loop {
+        let active_state = driver.is_active().await?;
+
+        let active_state_str = if active_state {
+            widget.state = State::Idle;
+            "active"
+        } else {
+            widget.state = State::Critical;
+            "inactive"
+        };
+
+        widget.set_values(map!(
+            "service" =>Value::text(config.service.clone()),
+            "active_state" =>Value::text(active_state_str.to_string()),
+        ));
+        api.set_widget(&widget).await?;
+
+        driver.wait_for_change().await?;
+    }
+}
+
+#[async_trait]
+trait Driver {
+    async fn is_active(&self) -> Result<bool>;
+    async fn wait_for_change(&mut self) -> Result<()>;
+}
+
+struct SystemdDriver {
+    proxy: UnitProxy<'static>,
+    active_state_changed: PropertyStream<'static, String>,
+}
+
+impl SystemdDriver {
+    async fn new(service: String) -> Result<Self> {
+        let dbus_conn = new_system_dbus_connection().await?;
+
+        let encoded_service = format!("{service}.service")
+            .chars()
+            .map(|c| {
+                if c.is_ascii_alphanumeric() {
+                    c.to_string()
+                } else {
+                    format!("_{:02x}", c as u8)
+                }
```

Comment:
Can you please add a comment explaining what is going on here? Also, are you sure about this `char as u8` cast? By the way, instead of `.collect::<Vec<String>>().join("")` you can do `.collect::<String>()`.